include/psa/crypto_struct.h

```diff
@@ -445,29 +452,18 @@ static inline psa_algorithm_t psa_get_key_algorithm(
     return( attributes->core.policy.alg );
 }
 
-/* This function is declared in crypto_extra.h, which comes after this
- * header file, but we need the function here, so repeat the declaration. */
-psa_status_t psa_set_key_domain_parameters(psa_key_attributes_t *attributes,
-                                           psa_key_type_t type,
-                                           const uint8_t *data,
-                                           size_t data_length);
-
 static inline void psa_set_key_type(psa_key_attributes_t *attributes,
                                     psa_key_type_t type)
 {
-    if( attributes->domain_parameters == NULL )
+    if( attributes->domain_parameters != NULL )
     {
-        /* Common case: quick path */
-        attributes->core.type = type;
-    }
-    else
-    {
-        /* Call the bigger function to free the old domain paramteres.
-         * Ignore any errors which may arise due to type requiring
-         * non-default domain parameters, since this function can't
-         * report errors. */
-        (void) psa_set_key_domain_parameters( attributes, type, NULL, 0 );
+        /* Free the old domain attributes if present */
+        mbedtls_free( attributes->domain_parameters );
```

Comment:
This change creates tighter coupling between “lightweight” attribute management in `crypto_struct.h` and domain parameter management which is too complex for inline functions defined in headers. Reducing `psa_set_key_domain_parameters(NULL)` to `free()` is correct with the current implementation, but is likely to fail when (if?) we implement custom Diffie-Hellman groups, which is the primary purpose of domain parameters: there's likely to be a cache of computed or allocated values which needs to garbage-collected in some way. This change is outright wrong if the domain parameters are stored in a crypto service which is a separate process from the application calling `psa_set_key_type`, and `psa_set_key_domain_parameters` makes an RPC call to the service.

Including some intelligence regarding which types have domain parameters in the header file would be ok. Currently RSA and FFDH do (or will when we get around to implementing it); ECC and anything symmetric don't.

Moving the domain parameter code out of `psa_crypto.c` into a smaller file would be ok, but I don't know if that would help you. I don't understand what the problem is: doesn't your build need to have `psa_crypto.o` anyway?

Reply:
No, it doesn't need `psa_crypto.o` with this change in place.

The setup is that we have moved our `MBEDTLS_XXX_ALT` drivers to sit on top of our PSA drivers, to avoid duplicate code when a library is built including both ALT and PSA drivers. This means the ALT drivers need to know about the PSA key attributes type.

The problem is that when a 'pure' mbedTLS library is built with these ALT drivers (because the specific user is not on PSA yet), `psa_crypto.o` is not included in that build. And IMO, shouldn't. That build contains mbed TLS sources + ALT driver sources + PSA driver sources only.

As far as our drivers go, we don't use RSA nor FFDH in this particular setup, so we could probably just reach into `attributes->core.type` from the ALT driver to set the key type instead of calling `psa_set_key_type`. But that is not the intended way to use the key attributes type, right?

As far as ideology goes, I find it strange that key domains are being touched at all when calling `set_key_type`. If one wants to re-use an attributes object, which evidently now contains dynamically allocated data, shouldn't there be a free-like operation instead?

Reply:
I am not sure to understand the point about client-server support. I have in mind that `psa_set_key_xyz()` are always used to set-up locally key attributes data structures and that key attributes are transferred to the server only when key attributes are `used` with a call to `psa_import_key()` or `psa_generate_key()` ... Do I have to change my mind?

Reply:
Most key attributes are small and fit into a fixed-size structure. This allows `psa_set_key_xxx` to be simple field assignments. Domain parameters don't fit this paradigm: they're big, so `psa_set_key_domain_parameters` might need to allocate memory, and therefore it can fail. In a client-server environment, it might need to allocate memory on the server side, because it's fairly common to use the same parameters for many keys, and therefore it be a reasonable optimization to only allocate memory once when importing many keys with the same parameters.

Domain parameters don't fit the mold of the other key attributes. That's part of why they're awkward.

Reply:
> As far as our drivers go, we don't use RSA nor FFDH in this particular setup, so we could probably just reach into attributes->core.type from the ALT driver to set the key type instead of calling psa_set_key_type. But that is not the intended way to use the key attributes type, right?

Please don't access fields of the attribute structure explicitly in your own code. We might change the layout at any time (for example to move things in/out of `core` or flatten `core` altogether).

Since you want a build without domain parameters, and your problem is caused by the existence of domain parameters, why don't you build with `static inline psa_status__t psa_set_key_domain_parameters(…) {return PSA_SUCCESS;}`?

> As far as ideology goes, I find it strange that key domains are being touched at all when calling set_key_type.

Domain parameters are only meaningful for a specific key type. They're a refinement of the key type that doesn't fit in the small, fixed-size field dedicated to the key type.

> If one wants to re-use an attributes object, which evidently now contains dynamically allocated data, shouldn't there be a free-like operation instead?

You can call `psa_set_xxx` multiple times on the same attributes object before using it (e.g. set defaults, then override some values). You can retrieve the attributes of a key and modify some of them to make a key with similar attributes to another key. There is a free-like operation (`psa_reset_key_attributes`), but that's only for when you want to erase all the attributes.

Reply:
In said client-server model, I fail to see how this would work given the current API. As @ronald-cron-arm mentioned, the client will still need to set up the key attributes structure locally before calling server APIs such as import or generate. Thus, it needs to pass the exact domain parameters to the server, always, on key generation/import, unless you're going to add some server-side domain parameter management API that allows creating entries for each domain and referencing them by ID across the server/client boundary?

Maybe the solution for key domain parameters in the key attributes is exactly this, an explicit resource management interface for domains?

Either way, that would be a much larger rework (and need spec'ing too, probably). For this specific PR, I note that I have explained the use case, and there were no alternatives suggested. So can we please move ahead here, and do the domain parameter support when it has been fully designed? The PR keeps status-quo with the current implementation, so I'm not sure why this should escalate in a full-on design discussion on custom domain support.

Reply:
Whether domain parameters are stored on the client or the server is (currently) implementation-specific. What is desirable depends a lot on where you have the memory for it and on how often you expect to reuse the same parameters.

I still fail to see the problem with keeping `psa_set_key_type` as is and overriding `psa_set_key_domain_parameters`.

Reply:
> I still fail to see the problem with keeping psa_set_key_type as is and overriding psa_set_key_domain_parameters.

Because I assume it'll trigger a linker error (duplicate definition) when the same setup is built with `psa_crypto.c` present?

Reply:
Also, how would this type of override be better than reaching into the struct? Both rely on unspecified library internals.

Reply:
After discussion, we propose to move `psa_set_key_domain_parameters`, `psa_get_key_domain_parameters` and `psa_reset_key_attributes` under a separate compilation option (likely with a separate `.c` file) so that it's possible to compile code that only accesses attributes but otherwise works without PSA crypto. This would be useful both for driver code, and for client code.